src/psyclone/tests/psyir/frontend/fparser2_directive_test.py

```diff
@@ -0,0 +1,209 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2021-2025, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author A. B. G. Chalk, STFC Daresbury Lab
+
+"""Performs pytest tests on the support for directives in the fparser2
+PSyIR front-end"""
+
+import pytest
+
+from psyclone.psyir.frontend.fortran import FortranReader
+from psyclone.psyir.nodes import (
+    CodeBlock,
+)
+
+
+def test_directive_after_decls():
+    """Test that the FortranReader correctly finds a directive immediately
+    after the declarations"""
+
+    code = """subroutine x()
+    integer :: i
+    !$omp barrier
+    i = 3
+    end subroutine"""
+    reader = FortranReader(ignore_comments=False, ignore_directives=False)
+    psyir = reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    # The directive is a codeblock
+    assert isinstance(routine.children[0], CodeBlock)
+    assert routine.children[0].debug_string() == "!$omp barrier\n"
+
+
+def test_directive_in_decls():
+    """Test that the FortranReader can handle a directive inside the
+    declarations"""
+    code = """subroutine x()
+    !$omp firstprivate
+    integer, dimension(100) :: i !dir$ aligned
+    i = 1
```

Comment:
I didn't think about this in the fparser PR but what happens if we have an inlined comment that looks like a directive.

e.g. `a = a + 1   !$omp atomic`

This is just a comment, but I am worried that fparser converts it to a preceding comment and then to a valid Directive. We need a test for this.